### PR TITLE
Added split_subparser.cpp and supporting Makefile changes.

### PR DESCRIPTION
**Pull request by dsyeager-zz**
Subparser and subbuffer are part of the json_parser repo at
https://github.com/VerizonDigital/json_parser.

Subparser approaches the problem a little differently so
I thought it was worth adding a test for it.

### CODE REVIEW

**Comment by ArashPartow** [comment]
The new numbers from Scott Yeager look more than a _bit off_.

**eg**:
1. split5.py was 71.1sec but is now 31.2 (now using an older less efficient python and note that both CPUs run at the same clock speed and nothing should be using vectorisations)
2. I can't replicate the difference between split_subparser and split7, in fact on 3 different systems it is slower than split7
3. split_subparser runs about the same as split6 beating it ~75% of time - codewise they are identical

It would be nice if these differences could be explained - and if more than one person can note the same set of results.

If an older version of python can outperform, by a margin of 2x, a newer version, wouldn't that imply a serious regression and shouldn't that be lodged as a bug report to the python people?

I know a lot of people that would give an arm and a leg for that kind of improvement in the python world.

In short something is dodgy with the system running the benchmark - Torbjörn it's best if you actually run the benchmarks on your AMD Athlon, at least that way all new additions will be benchmarked with a common base line architecture .

**Comment by tobbez** [comment]
Regarding your first point, we can probably assume some other optimization is in play, since the numbers for Python 2.7.6 seem to be in line with the ones for Python 2.6.9 (on the same machine):

```
=== System info
Linux 3.11.6-gentoo x86_64 GNU/Linux
AMD Athlon(tm) 64 X2 Dual Core Processor 4200+
g++ (Gentoo 4.8.4 p1.0, pie-0.6.1) 4.8.4
Python 2.7.6
=== End System info

./split.py         Python: Saw 20000000 lines (60000000 words/806116396 chars) in 60.3 seconds.  Crunch Speed: 331918.2
./split5.py        Python: Saw 20000000 lines (60000000 words/806116396 chars) in 60.6 seconds.  Crunch Speed: 329800.9
./split1           C++   : Saw 20000000 lines (60000000 words/806116396 chars) in 26.9 seconds.  Crunch speed: 744782.6
./split2           C++   : Saw 20000000 lines (60000000 words/806116396 chars) in 39.3 seconds.  Crunch speed: 509313.6
./split6           C++   : Saw 20000000 lines (60000000 words/806116396 chars) in 5.6 seconds.  Crunch speed: 3594805.3
./split7           C++   : Saw 20000000 lines (60000000 words/806116396 chars) in 5.3 seconds.  Crunch speed: 3793841.5
./split8           C++   : Saw 20000000 lines (60000000 words/806116396 chars) in 64.1 seconds.  Crunch speed: 312089.1
./split9           C++   : Saw 20000000 lines (60000000 words/806116396 chars) in 34.3 seconds.  Crunch speed: 583320.4
./split_subparser  C++   : Saw 20000000 lines (60000000 words/806116396 chars) in 4.2 seconds.  Crunch speed: 4780442.1
./splitc1          C++   : Saw 20000000 lines (60000000 words/806116396 chars) in 17.8 seconds.  Crunch speed: 1123487.6
./splitc2          C++   : Saw 20000000 lines (60000000 words/806116396 chars) in 17.3 seconds.  Crunch speed: 1156258.1
./splitc3          C++   : Saw 20000000 lines (60000000 words/806116396 chars) in 14.3 seconds.  Crunch speed: 1396034.4
```

As for your second point, on both of the two systems where I ran the tests, split_subparser was slightly faster than split7.
System 1:

```
=== System info
Linux 3.11.6-gentoo x86_64 GNU/Linux
AMD Athlon(tm) 64 X2 Dual Core Processor 4200+
g++ (Gentoo 4.8.4 p1.0, pie-0.6.1) 4.8.4
Python 2.7.10
=== End System info

./split.py         Python: Saw 20000000 lines (60000000 words/806116396 chars) in 69.1 seconds.  Crunch Speed: 289348.7
./split5.py        Python: Saw 20000000 lines (60000000 words/806116396 chars) in 71.5 seconds.  Crunch Speed: 279661.8
./split1           C++   : Saw 20000000 lines (60000000 words/806116396 chars) in 26.3 seconds.  Crunch speed: 760336.6
./split2           C++   : Saw 20000000 lines (60000000 words/806116396 chars) in 39.3 seconds.  Crunch speed: 509548.8
./split6           C++   : Saw 20000000 lines (60000000 words/806116396 chars) in 5.6 seconds.  Crunch speed: 3571283.6
./split7           C++   : Saw 20000000 lines (60000000 words/806116396 chars) in 5.3 seconds.  Crunch speed: 3802013.6
./split8           C++   : Saw 20000000 lines (60000000 words/806116396 chars) in 66.1 seconds.  Crunch speed: 302663.9
./split9           C++   : Saw 20000000 lines (60000000 words/806116396 chars) in 35.8 seconds.  Crunch speed: 558671.2
./split_subparser  C++   : Saw 20000000 lines (60000000 words/806116396 chars) in 4.3 seconds.  Crunch speed: 4703532.7
./splitc1          C++   : Saw 20000000 lines (60000000 words/806116396 chars) in 18.0 seconds.  Crunch speed: 1111424.2
./splitc2          C++   : Saw 20000000 lines (60000000 words/806116396 chars) in 16.2 seconds.  Crunch speed: 1231258.4
./splitc3          C++   : Saw 20000000 lines (60000000 words/806116396 chars) in 13.9 seconds.  Crunch speed: 1440719.6
```

System 2:

```
=== System info
Linux 3.13.0-57-generic x86_64 GNU/Linux
Intel(R) Core(TM) i7 CPU         860  @ 2.80GHz
g++ (Ubuntu 4.8.2-19ubuntu1) 4.8.2
Python 2.7.6
=== End System info

./split.py         Python: Saw 20000000 lines (60000000 words/806116396 chars) in 31.6 seconds.  Crunch Speed: 632765.8
./split5.py        Python: Saw 20000000 lines (60000000 words/806116396 chars) in 25.0 seconds.  Crunch Speed: 800707.9
./split1           C++   : Saw 20000000 lines (60000000 words/806116396 chars) in 18.9 seconds.  Crunch speed: 1059989.0
./split2           C++   : Saw 20000000 lines (60000000 words/806116396 chars) in 18.7 seconds.  Crunch speed: 1071126.9
./split6           C++   : Saw 20000000 lines (60000000 words/806116396 chars) in 2.4 seconds.  Crunch speed: 8282628.5
./split7           C++   : Saw 20000000 lines (60000000 words/806116396 chars) in 2.3 seconds.  Crunch speed: 8792898.6
./split8           C++   : Saw 20000000 lines (60000000 words/806116396 chars) in 30.6 seconds.  Crunch speed: 653007.7
./split9           C++   : Saw 20000000 lines (60000000 words/806116396 chars) in 17.5 seconds.  Crunch speed: 1144533.8
./split_subparser  C++   : Saw 20000000 lines (60000000 words/806116396 chars) in 2.0 seconds.  Crunch speed: 10000062.9
./splitc1          C++   : Saw 20000000 lines (60000000 words/806116396 chars) in 11.2 seconds.  Crunch speed: 1782272.9
./splitc2          C++   : Saw 20000000 lines (60000000 words/806116396 chars) in 9.7 seconds.  Crunch speed: 2060599.1
./splitc3          C++   : Saw 20000000 lines (60000000 words/806116396 chars) in 8.6 seconds.  Crunch speed: 2338316.4
```

**Comment by dsyeager-zz** [comment]
Gents,
    The two tests that Torbjörn did on his "AMD Athlon(tm) 64 X2 Dual  
Core Processor 4200+" show a lot of variance in the python tests. His  
second test system  (I7 2.8GZ) is faster than my python results. The  
variability of the python results coupled with the consistent disk  
access speeds could explain a lot of the "dodgy" results.

Scott

Quoting Torbjörn Lönnemark notifications@github.com:

**Comment by ArashPartow** [comment]
@Torbjörn, @Scott,

wrt disk access, that is definitely an issue and can affect timings, however from analysis on my system, by the time split2 is executed the file has completely been paged into memory - I'm running with 64GB RAM, which should greatly reduce any hdd seek related distortions.

On the topic of python, it's very likely the better l1/l2 cache on the i7 is doing something there - or there is some kind of regression as even a significant difference in the calibre of the l1/l2 between the athlon and the i7 should not result in that much of a difference.

That said, I'll accept the results as is, and put it down to a combination of superior implementation from Scott, l1/l2 cache improvements and gremlins of a temporaneous nature. :)
